### ISSUE TITLE
Refactor generate_stats_figs.R

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#32]
Refactor existing code:

* Replace use of `--dir_data` with command-line arguments. The script currently works as follows:
  - develop branch: uses a `--dir_data=data/` argument to access and read `data/yeast_codon_pos_i200.RData` and `/data/yeast_tRNAs.tsv`.
  - master branch: assumes these files are peers of the script in `scripts/`.
* Pass these file names as command-line arguments to make the dependency explicit.

Conform to R styles.

Provide complementary documentation.

Provide complementary regression and unit tests.

Alternatively do a complete rewrite using [tidyverse](https://www.tidyverse.org/)

**Comment by kavousan** [comment]
This is a bit untidy now, but we have made progress and learnt a lot. We treat this as a background task for December, in the knowledge that we will need this in a better state before we can visualise the new datasets.

**Comment by ewallace** [comment]
Note that to debug an R script including command-line arguments, we have to replace 
`Rscript` with `R --no-save --no-restore --args`. Then to run a script, source it from the R command line, for example `source("rscripts/generate_stats_figs.R")`. This means that if there's R debugging, for example a call to `browser()`, then it will debug from the R command line, with the arguments supplied.

For example, in the vignette, we should replace
```Rscript --vanilla rscripts/generate_stats_figs.R --Ncores=1 --MinReadLen=10 --MaxReadLen=50 --Buffer=250 --PrimaryID=Name --dataset=vignette --hdFile=vignette/output/WT3AT.h5 --out_prefix=vignette/output/WT3AT --orf_fasta=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --dir_out=vignette/output --do_pos_sp_nt_freq=True --t_rna=data/yeast_tRNAs.tsv --codon_pos=data/yeast_codon_pos_i200.RData --features_file=data/yeast_features.tsv --orf_gff_file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --count_threshold=64
```
with
```R --no-save --no-restore --vanilla --args --Ncores=1 --MinReadLen=10 --MaxReadLen=50 --Buffer=250 --PrimaryID=Name --dataset=vignette --hdFile=vignette/output/WT3AT.h5 --out_prefix=vignette/output/WT3AT --orf_fasta=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --dir_out=vignette/output --do_pos_sp_nt_freq=True --t_rna=data/yeast_tRNAs.tsv --codon_pos=data/yeast_codon_pos_i200.RData --features_file=data/yeast_features.tsv --orf_gff_file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --count_threshold=64```

Where in the developer guide does this info need to go?

**Comment by mikej888** [comment]
A new section should be added, perhaps after the "Logging" section.

**Comment by ewallace** [comment]
Suggestions for refactoring with tidyverse approach:

* use `purrr` functions such as `map` to replace as many awful loops as possible
* for example, can replace `colSums` on matrix by `purrr::reduce(.,`+`)` on list of vectors
* replace `BiocGenerics::Reduce` with `purrr::reduce`
* replace `gff` and BiocGenerics with tibble `gff_df` and tidy functions
* continue to replace generic variable names such as `out` and `data_mat` with informative names

**Comment by ewallace** [comment]
Section added to developer guide in `read-frame-89` branch.

**Comment by ewallace** [comment]
Suggestion from @lianafaye to split `generate_stats_figs.R` into one script *before* a-site assignment, then another script *assuming a-site assignment rules known*. The first script could then output an estimated a-site offset file, i.e. #104.

**Comment by mikej888** [comment]
Addressing the following tickets has introduced changes to generate_stats_figs.R (all present in branch [provenance-90-115-123](https://github.com/riboviz/RiboViz/tree/provenance-90-115-123) 

* Rename config.yaml variables for consistency #86 - command-line options, and associated variables were renamed; also output_dir is now used.
* Remove sample prefix from file names #114
* Add git commit hash etc to log #90 - helper function in provenance.R is called to get version info which is printed.
* Insert provenance comments into tsv files #115- helper function in provenance.R is called to write comment header to all TSV files.

These will create a number of conflicts when merging provenance-90-115-123 with regen-stats-figs-32. We can decide how to synch this in the new year.

**Comment by ewallace** [comment]
Commit b9fe9cd includes many refactoring steps:
* placed all functions from `generate_stats_figs.R` into separate script `read_count_functions.R`
* `GetGene...` functions consistently start with arguments `(gene,dataset,hdf5file,...)`
* inserted comments and TODOs for more refactoring
* Check for 3nt periodicity globally / gene_pos_counts now thoroughly tidyverse
* moved tidy gene_sp_read_length to GetGeneReadLength
* replace gff IRanges call with gff_df tidy data frame throughout
* Position specific distribution of reads / reads_per_codon_etc thoroughly tidyverse
* merged #114
* Merged 9ce97e4 #115 / config files

Next priorities:
* Tidy: Biases in nucleotide composition along mapped read lengths / do_pos_sp_nt_freq
* Refactor: Codon-specific ribosome densities, to separate calculation of normalized densities with comparison to trnas or other codon-specific features.
* need to rename `reads_per_codon_etc` to `reads_per_codonpos` or something more informative.

**Comment by ewallace** [comment]
To be picked up by @FlicAnderson.

Suggestion is to first:
* Merge `develop` into `regen-stats-figs-32` and test on vignette.
* Review @ewallace 's updates from b9fe9cd, report any that don't make sense.
* style the R code as developer guide, fix glaring problems e.g. CamelCase or not
* suggest/implement any other places where repeated code should be collected to a separate function in `read_count_functions.R`

Then probably (mostly as previous message):
* Tidy: Biases in nucleotide composition along mapped read lengths / do_pos_sp_nt_freq
* Refactor: Codon-specific ribosome densities, to separate calculation of normalized densities with comparison to trnas or other codon-specific features.
* Rename reads_per_codon_etc to reads_per_codonpos or something more informative.
* Make changes suggested in #106 

Goal of complete refactored and tidy `generate_stats_figs.R` script.

*After* that, goal is to port the remaining code into a `generate_stats_figs.Rmd` that creates a single html report. This would be run from the command line in `prep_riboviz.py`, with the same inputs, and output to a single html file in the sample directory.

**Comment by FlicAnderson** [comment]
See relevant issue #134 for issue breaking interactive R behaviours (get_Rscript_filename() at fault)

**Comment by FlicAnderson** [comment]
#### Status update: 
General refactoring still ongoing, after delay caused by issue #134 (now resolved); new potential issue picked up by regression test in different file outputs for `pos_sp_rof_norm_reads`  from `generate-stats-figs.R`

:champagne:  Integrated fix #134 contents via latest `develop` into `regen-stats-32` branch; now able to debug / run interactively. 

:cry: Downside: running regression test with a reasonably recent set of example files and getting failures. With `pos_sp_rof_norm_reads` .tsv outputs differing between `develop` and `regen-stats-32`.   

I believe these aren't due to the new merges-in, I'm pretty certain they maybe relate to refactoring changes on branch `regen-stats-32` that Edward was working on in December. 

The regression test fails twice (ie. the two vignette samples both fail to match their regression test example equivalents) on `generate_stats_figs.R`.

There's a clear explanation for some differences to be expected due to `regen-stats-32` including rounding to 4 significant figures in the writing out to tsv stage in this branch vs develop (refactoring changes made in December 2019). 

BUT, even despite that, the differences in means and SD seem a question of being different numbers, rather than diffences being as a result of differing numbers of digits.

Investigating the differences in the .tsv files: 

/home/flic/vignette-20200219-3f928a/output/WT3AT (below) VIA DEVELOP
```
# Created by: RiboViz			
# Date: 2020-02-19 13:34:28			
# File: rscripts/generate_stats_figs.R			
# Version: commit 3f928a4fbe6b3597ffa867f8499654526060cf74 date 2020-02-19 11:46:12			
Position	Mean	                   SD	                       End
1	0.114900896763824	0.114900896763824	5'
2	1.75191797936548	       1.01523280394192	      5'
3	0.314190729667586	0.212850704292156	5'
4	0.181330841065078	0.122780862570594	5'
```
/home/flic/RiboViz/vignette/output/WT3AT (below) VIA REGEN-STATS-32
```
# Created by: RiboViz			
# Date: 2020-02-25 17:01:09			
# File: /home/flic/RiboViz/rscripts/generate_stats_figs.R			
# Version: commit ae66f197d6f7a027c13ac67f5e5c439bfef344c9 date 2020-02-25 16:53:30			
Position	Mean	SD	End
1	           0	       0	5'
2	           1	    1.414	5'
3	           1.5	3	5'
4	           0.5	1	5'
```
(forgive the random alignment)

These numbers seem fairly different (compare Mean in row 2 & 3 between sets) , but this might (?tentative possible explanation follows?) be accounted for by the different methods of creating the `pos_sp_rof_norm_reads` between `develop` and the refactored `regen-stats-32` code?

@Develop (creates pos_sp_rpf_norm_reads in one go)
```R
 # Create a dataframe to store the output for plots/analyses
  pos_sp_rpf_norm_reads <- data.frame(
    Position = c(1:500, 0:-499),
    Mean = c(m5p, m3p),
    SD = c(s5p, s3p),
    End = factor(rep(c("5'", "3'"), each = 500), levels = c("5'", "3'"))
  )

  # Plot
  # ...

  # Save plot and file
  ggsave(pos_sp_rpf_norm_reads_plot, filename = file.path(output_dir, paste0(output_prefix, "pos_sp_rpf_norm_reads.pdf")))
  tsv_file_path <- file.path(output_dir, paste0(output_prefix, "pos_sp_rpf_norm_reads.tsv"))
  write_provenance_header(this_script, tsv_file_path)
  write.table(
    pos_sp_rpf_norm_reads,
    file = tsv_file_path,
    append = T,
    sep = "\t",
    row = F,
    col = T,
    quote = F
  )
}
```

@regen-stats-32 (creates 5' and 3' separately, joins into pos_sp_rpf_norm_reads)
```R
pos_sp_rpf_norm_reads_5end <- 
    reads_per_codon_etc %>%
    # select genes over count_threshold and length_threshold
    filter(SumAsiteCt > count_threshold, LengthCodons > length_threshold) %>%
    # Extract from codon positions 1 to length_threshold
    transmute(NormCtperCodon5end = 
                map(CountPerCodon,
                    ~extract(.,seq_len(length_threshold)))) %>%
    # remove NormCts as vector and (3 lines later) make these rows of a matrix
    dplyr::pull(NormCtperCodon5end) %>%
    unlist %>% 
    matrix(byrow=T,ncol=length_threshold) %>%
    # calculate column means and standard deviations in tidy data frame
    {tibble(Position = seq_len(length_threshold),
            Mean = colMeans(.) %>% as.numeric,
            SD   = apply(.,2,sd),
            End  = "5'")}
  
  # metagene mean normalized read counts for 3' end stop-length_threshold to stop
  pos_sp_rpf_norm_reads_3end <- 
    # ... same method as above, but for 3' end
  
  pos_sp_rpf_norm_reads <- bind_rows(pos_sp_rpf_norm_reads_5end,
                                     pos_sp_rpf_norm_reads_3end)
  
  # Plot
 # ... 

  # Save plot and file
  ggsave(pos_sp_rpf_norm_reads_plot, filename = file.path(output_dir, paste0(output_prefix, "pos_sp_rpf_norm_reads.pdf")))
  tsv_file_path <- file.path(output_dir, paste0(output_prefix, "pos_sp_rpf_norm_reads.tsv"))
  write_provenance_header(path_to_this_script, tsv_file_path)
  write.table(
    pos_sp_rpf_norm_reads %>%
      # mutate_if() mutates if the data is numeric, rounds values to specified number digits
      mutate_if(is.numeric, signif, digits=4), 
    file = tsv_file_path,
    append = T,
    sep = "\t",
    row = F,
    col = T,
    quote = F
  )
```
I've removed a bit of unnecessary code from these examples (plotting method, similar codeblocks), but left in the write-out (save plot and file) code since that includes the `mutate_if()` line which creates the significant figure differences.

@ewallace any thoughts/ideas on how to best troubleshoot or test these differences?  

I've pushed branch `regen-stats-32`[commit 0ad0b12](https://github.com/riboviz/RiboViz/commit/0ad0b12bd6da5bc43dd53706d4098762a984d126) despite the regression test failure, but this is still definitely ongoing work. 

Otherwise, to re-iterate, general refactoring still ongoing, after delay caused by issue #134

**Comment by FlicAnderson** [comment]
Refactoring Plan: 

After discussion with @mikej888, decided to take the following approach with code at `regen-stats-32` with regression testing** using existing regression tests throughout: 

1) Coarse-grained decomposition
- Split out the large independent chunks of code & isolate these as functions
  - interactive & option_list setup portions (leave alone currently, potentially in future split off as functions or separate script?)
  - 3nt periodicity
  - Lengths all mapped reads
  - Biases in nucleotide composition
  - Calculate read frame for every orf
  - RPF position specific distribution of reads
  - mRNA position specific distribution of reads
  - TPMs of genes
  - TPMs correlations with features
(2 DAYS)

2) Re-organising to put computation code first, then outputs (ie .pdf; .tsv) last
(1 DAY)

3) Style for consistency to follow [Google R Style Guide]( https://google.github.io/styleguide/Rguide.html). 
Starting from the end of the script & working back to more evenly spread the refactoring effort across generate_stats_figs.R, since previous approaches started at the start and worked forward.
Key points: 
  - Google prefers identifying functions with `BigCamelCase` to clearly distinguish them from other objects
- Don’t use `attach()`
- Use explicit returns 
- Qualify namespaces (e.g. use `purrr::map()` rather than `map()` )
- No right-hand assignment (e.g. always `foo <-`, never `-> foo`)
(1.5 WEEK)

4) Finer-grained decomposition within these larger chunks
  - Splitting out more functions (keeping 'testibility' in mind)
(1.5 WEEKS)
   
5) Move functions out to separate script (ie. read_count_functions.R) where this makes sense
(2 DAYS)

6) Add more testing for each new function: e.g. are inputs as expected, is data structure as expected, are column names / key values as expected
(ESTIMATE LATER)


** = using current 2-fails output as 'expected' dataset [see more info above](https://github.com/riboviz/riboviz/issues/32#issuecomment-590991487) to compare against, until code is refactored sufficiently & THEN can investigate the reasons behind those regression tests fails compared to develop/master.

**Comment by mikej888** [comment]
This could be followed by supporting RMarkdown export of data.

**Comment by mikej888** [comment]
Edward suggests:

* Coarse-grained decomposition including documentation of functions esp. inputs and outputs and their expected/actual contents and move these out to a separate script.
* Re-organising within those large functions to put computation code first, then outputs (ie .pdf; .tsv) last and decompose into "compute" and "plot" functions.
* Replace PDF export with RMarkdown export.
* Implement unit tests for scientific validity of the "compute" functions e.g. are inputs as expected, is data structure as expected, are column names / key values as expected

... then...

* Style for consistency to follow Google R Style Guide.
* Finer-grained decomposition within the larger functions.
* Move functions out to separate script (ie. `read_count_functions.R`) where this makes sense

**Comment by FlicAnderson** [comment]
Updates on refactoring: 

Working through generate_stats_figs.R as planned [previously](https://github.com/riboviz/riboviz/issues/32#issuecomment-597745061) and breaking code up into big chunks, and within that into medium chunks.

LARGE CHUNKS:
 3nt periodicity
 Lengths all mapped reads
 Biases in nucleotide composition
 Calculate read frame for every orf
 RPF position specific distribution of reads
 mRNA position specific distribution of reads
 TPMs of genes
 TPMs correlations with features

MEDIUM CHUNKS:
Following this naming convention:
 CalculateFunction - computational code
 PlotFunction - plot creation
 SaveFunction - PDF writeouts of plots
 WriteFunction - .TSV file creation

Keeping the medium-size function chunks in generate_stats_figs.R for now, for ease of working, and will then move these out to external script (such as read_counts_functions.R, which may be renamed, or to a variety of appropriately named .R scripts).

**Comment by FlicAnderson** [comment]
Regression testing against the '2-fail' files as planned previously to allow refactoring so I can then investigate the issue and confirm what the output should be and if we need to update the regression test files (and to what), later in the proceedings.

**Comment by FlicAnderson** [comment]
I'm bearing in mind (after a discussion with @ewallace on 24th March) that the Save* (PDF writeout of plots) medium-size functions won't be necessary in future once the script is outputting plots as .html as part of an .Rmd, but as removing these currently would cause additional (missing .pdf file) regression test fails, I think it's best to keep these in place for now.

**Comment by FlicAnderson** [comment]
I still need to address the problem caused by .h5 files not playing well with .Rmd as this impacts the plan to replace the .PDF output plots with an html output, and also causes problems for Ania currently as she's writing her code.  :question: Worth me creating this as a separate issue, I think?

**Comment by FlicAnderson** [comment]
One thing I noticed while running regression tests while refactoring is that I can sometimes get failures caused by having a connection open to an .h5 file in Rstudio/etc, while trying to run regression test in terminal. This should probably have been obvious but I'm just putting it here in case: a) I forget, or b) someone else tries and gets the same issue. 

From `riboviz.20200324.184032.log`:
```console
2020-03-24 18:40:32,140:riboviz.tools.prep_riboviz:INFO: Created by: RiboViz Date: 2020-03-24 18:40:32.134990 Command-line tool: /home/flic/miniconda3/bin/pytest File: /home/flic/riboviz/riboviz/tools/prep_riboviz.py Version: commit 3a8775600b8aa8b213939f61404b329fb2b416e2 date 2020-03-24 18:35:47+00:00 
# ...
2020-03-24 18:40:47,520:riboviz.tools.prep_riboviz:ERROR: Problem processing sample: WTnone
2020-03-24 18:40:47,521:riboviz.tools.prep_riboviz:ERROR: AssertionError
Traceback (most recent call last):
  File "/home/flic/riboviz/riboviz/tools/prep_riboviz.py", line 515, in process_samples
    sample_logs_dir, run_config)
  File "/home/flic/riboviz/riboviz/tools/prep_riboviz.py", line 430, in process_sample
    orf_gff_file, config, log_file, run_config)
  File "/home/flic/riboviz/riboviz/workflow.py", line 533, in bam_to_h5
    run_config.is_dry_run)
  File "/home/flic/riboviz/riboviz/process_utils.py", line 130, in run_logged_command
    run_command(cmd, f, f)
  File "/home/flic/riboviz/riboviz/process_utils.py", line 22, in run_command
    assert exit_code == 0, "%s failed with exit code %d" % (cmd, exit_code)
AssertionError: ['Rscript', '--vanilla', '/home/flic/riboviz/rscripts/bam_to_h5.R', '--num-processes=1', '--min-read-length=10', '--max-read-length=50', '--buffer=250', '--primary-id=Name', '--secondary-id=NULL', '--dataset=vignette', '--bam-file=vignette/output/WTnone/WTnone.bam', '--hd-file=vignette/output/WTnone/WTnone.h5', '--orf-gff-file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3', '--is-riboviz-gff=True', '--stop-in-cds=False'] failed with exit code 1
```
Snippet from regression test output of `TPMs_collated.tsv` tests: 
```console
riboviz/test/regression/test_vignette.py:347: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
riboviz/compare_files.py:58: in compare_files
    utils.equal_tsv(file1, file2)
riboviz/utils.py:197: in equal_tsv
    equal_dataframes(data1, data2, tolerance)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data1 =           ORF    WTnone     WT3AT
0     YAL001C     116.7     953.0
1     YAL002W     212.8     365.8
2     YAL003W  1... YAL067C       0.0     290.4
66  YAL067W-A       0.0    1287.7
67    YAL068C       0.0     431.9

[68 rows x 3 columns]
data2 =           ORF     WT3AT
0     YAL001C     953.0
1     YAL002W     365.8
2     YAL003W  100458.6
3     YAL005C  239069....
64    YAL065C     408.0
65    YAL067C     290.4
66  YAL067W-A    1287.7
67    YAL068C     431.9

[68 rows x 2 columns]
tolerance = 0.0001
```
`WTnone` sample is missing in data2.  This was the sample I'd had loaded and was tinkering with in Rstudio.

Fix: 
I closed Rstudio after clearing the environment, then re-ran the regression test, and all was well again. :+1:

**Comment by ewallace** [comment]
Isn't this h5 issue is the same as #106

**Comment by FlicAnderson** [comment]
Update on how I'm going to resolve .h5 issue in new branch (`h5-open-106`) instead of within #32 branch as I've started off doing > https://github.com/riboviz/riboviz/issues/106#issuecomment-610955742

**Comment by FlicAnderson** [comment]
Progress update: 

LARGE CHUNKS: 
3nt periodicity - ThreeNucleotidePeriodicity() done; medium chunks isolated out.
Lengths all mapped reads - DistributionOfLengthsMappedReads() done; medium chunks isolated out.
Biases in nucleotide composition - BiasesInNucleotideCompositionAlongMappedReadLengths() done; medium chunks IN PROGRESS.
Calculate read frame for every orf - NEXT TARGET

TO DO (all levels)
RPF position specific distribution of reads
mRNA position specific distribution of reads
TPMs of genes
TPMs correlations with features

**Comment by FlicAnderson** [comment]
Shifted to focus on resolving #106 in branch `h5-open-106` before shortly returning to refactoring into large then medium chunks. 

Progress through chunks still [as above](https://github.com/riboviz/riboviz/issues/32#issuecomment-610988201).

**Comment by kavousan** [comment]
Aiming for @ewallace to be able to review the code by 1 May.

**Comment by FlicAnderson** [comment]
Completed splitting out large chunks as follows:

- 3nt periodicity
- Lengths all mapped reads
- Biases in nucleotide composition
- Calculate read frame for every orf
- RPF position specific distribution of reads
- mRNA position specific distribution of reads
- TPMs of genes
- TPMs correlations with features

Split out medium chunks according to naming convention: 

- ReadFunction - read in data where necessary
- CalculateFunction - computational code
- PlotFunction - plot creation
- SaveFunction - PDF writeouts of plots
- WriteFunction - .TSV file creation

To do: 
- address code in medium chunks (especially bottom half of script)
- look at whether I need `return(output)` at the end of functions, then still need to declare `output <- function(inputs)` when using the function?
- style assessment
- .Rmd-ify (but in different branch, as it's a different issue!)

**Comment by FlicAnderson** [comment]
Figure out regression failure: run develop and regen-stats-32 on REAL data, compare plots/outputs for Wednesday's dev meeting.

**Comment by FlicAnderson** [comment]
Onward plans:

Split out medium chunks next into new file.

Then have .Rmd running the big chunks.

**Comment by FlicAnderson** [comment]
> Onward plans:
> 
> Split out medium chunks next into new file.
> 
> Then have .Rmd running the big chunks.

Medium chunk removal to new file mentioned in #153 [here](https://github.com/riboviz/riboviz/issues/153#issuecomment-622403999)

**Comment by FlicAnderson** [comment]
Further advice on good-to-have improvements / things to address in future or subsequent issues for `regen-stats-32` from Edward which were amongst my notes:
  - potentially replace/tweak plot_ribogrid() to follow StyleGuide / merge w/ PlotStartCodonRiboGrid
  - list which medium-level functions use purrr, which use lapply, etc
  - be aware for future: the SAVE functions won't be in ~Plot_functions() as we'll be writing to .Rmd file.
  - try and go through the code and look at the data structures and think about WHAT IS BEING REPRESENTED! 
  - Try to plot individual bits of that data, and plot those intermediate steps.
  - When testing/refactoring, squeeze the data into like only 3 genes' worth and keep it manageable.
  - prioritise the worst bits; THEN regression tests (probably needs real data); THEN purrr instead of lapply.
  - come up with a proposal on purrr instead of lapply, read up, come up with a specification.
  - Go through this and make list of the tests we need. check inputs are in right format, check functions do the right thing.

**Comment by FlicAnderson** [comment]
Regression Test Fails

Currently the regression tests using vignette fail for 2 files, as mentioned [here](https://github.com/riboviz/riboviz/issues/32#issuecomment-590991487). 

> There's a clear explanation for some differences to be expected due to regen-stats-32 including rounding to 4 significant figures in the writing out to tsv stage in this branch vs develop (refactoring changes made in December 2019).

> BUT, even despite that, the differences in means and SD seem a question of being different numbers, rather than diffences being as a result of differing numbers of digits.

@ewallace mentioned that a key difference between the two sets of code are that the `regen-stats-32` version filters out reads which are outside certain threshholds: 
 `# select genes over count_threshold and length_threshold`
... which would mean that the numbers should be expected to be different and this new code is likely to be sensible.  This needs to be checked.

* Branch `regen-stats-32` tested by running full-size data from 3 samples of G-Sc_2014 through riboviz.  Run completed successfully. 

Need to run same samples of G-Sc_2014 through `develop` branch to compare outputs in regression-test style, to check differences reflect those shown by the regression test (either [this dated one](https://github.com/riboviz/regression-test-data-20200221) or [the 2.0.beta one](https://github.com/riboviz/regression-test-data-2.0.beta)).  This is potentially a good use for #155 code?

Output plots for the `pos_sp_rpf_norm_reads` code section for comparison to develop outputs when run:
WTnone `regen-stats-32` version:
[pos_sp_rpf_norm_reads.pdf](https://github.com/riboviz/riboviz/files/4587511/pos_sp_rpf_norm_reads.pdf)

Develop versions: will be run after meeting and compared.  Worth setting a time Thursday/Friday to discuss these @ewallace ?

**Comment by kavousan** [comment]
@FlicAnderson to try to roll back the Dec 2019 changes and see if we can then pass the regression tests.

**Comment by FlicAnderson** [comment]
Latest changes: 

- DONE: rolled-back sections passing regression test in `regen-stats-32` vs `develop` branch & running vignette fine
- STILL TO DO: run on full-size data; compare to same outputs in develop via #155  (and this will help review/close pull request #161 ) then **create pull request**
- STILL TO DO:  Create issue to create test for and (perhaps as separate issue) investigate differences between rolled-back code ([these changes, effectively](https://github.com/riboviz/riboviz/commit/4967c1e5401b773422722ab57544ae3814daa872) in these sections:
  - Position specific distribution of reads - RPF method
  - Position specific distribution of reads - mRNA dataset method
  - Codon-specific ribosome densities for correlations with tRNAs

**Comment by FlicAnderson** [comment]
Also STILL TO DO: should include before pull request: additions to documentation following #169 as per Mike's suggestions [here](https://github.com/riboviz/riboviz/issues/169#issue-605488208)

**Comment by FlicAnderson** [comment]
DONE: run on full-size data; compare to same outputs in develop via #155 (reviewed/merged pull request #161)

Merging issue #155 / PR #161 into develop let @mikej888 merge a whole bundle of dependent pull requests into develop, which I've now merged into `regen-stats-32`.  

Vignette runs, passes regression tests, now running on full-size data (don't expect any issues, since the code of #32 has been pretty thoroughly tested by now, but running again to ensure I can compare all the interactions with new develop code) and will compare to a new full-size run of latest `develop` using the newly generalised regression test command:
```
pytest -v riboviz/test/regression/test_regression.py --expected=DIRECTORY --skip-workflow --config-file=CONFIGFILE
```
At that point, I'll pull request and it's up for review; aiming to get this done by 1600 dev meeting today.

**Comment by FlicAnderson** [comment]
Full-size 1 sample dataset (G-Sc_2014 WTnone sample) run on `regen-stats-32` branch (last commit 3eec9037e3), ran well. Compared with full-size same sample run on `develop` branch (last commit 10f18aaed31da8039a7f2cfc823aa40e47b0cefd) using the generalised regression test - full match!  

Very confident that regen-stats-32 code output matches develop as of commit 8628d7ea47bf8c35cbae3a9e6b9fd4ee7650ef42, so ready to pull request (have merged latest develop at 10f18aaed, includes nextflow branch content).

Pull-requesting, and looking forward to review for any improvements I can make.

**Comment by FlicAnderson** [comment]
Pull request submitted: #184

There's probably something crucial I've forgotten, but this still feels like extreme progress :smile:

**Comment by kavousan** [comment]
@ewallace to review w/b 22 June 2020.
